cpp/src/arrow/type_test.cc

```diff
@@ -892,6 +892,150 @@ class TestUnifySchemas : public TestSchema {
           << lhs_field->ToString() << " vs " << rhs_field->ToString();
     }
   }
+
+  void CheckUnifyAsymmetric(
+      const std::shared_ptr<Field>& field1, const std::shared_ptr<Field>& field2,
+      const std::shared_ptr<Field>& expected,
+      const Field::MergeOptions& options = Field::MergeOptions::Defaults()) {
+    ARROW_SCOPED_TRACE("options: ", options);
+    ARROW_SCOPED_TRACE("field2: ", field2->ToString());
+    ARROW_SCOPED_TRACE("field1: ", field1->ToString());
+    ASSERT_OK_AND_ASSIGN(auto merged, field1->MergeWith(field2, options));
+    AssertFieldEqual(merged, expected);
+  }
+
+  void CheckUnify(const std::shared_ptr<Field>& field1,
+                  const std::shared_ptr<Field>& field2,
+                  const std::shared_ptr<Field>& expected,
+                  const Field::MergeOptions& options = Field::MergeOptions::Defaults()) {
+    CheckUnifyAsymmetric(field1, field2, expected, options);
+    CheckUnifyAsymmetric(field2, field1, expected, options);
+  }
+
+  void CheckUnifyFailsInvalid(
+      const std::shared_ptr<Field>& field1, const std::shared_ptr<Field>& field2,
+      const Field::MergeOptions& options = Field::MergeOptions::Defaults()) {
+    ARROW_SCOPED_TRACE("options: ", options);
+    ARROW_SCOPED_TRACE("field2: ", field2->ToString());
+    ARROW_SCOPED_TRACE("field1: ", field1->ToString());
+    ASSERT_RAISES(Invalid, field1->MergeWith(field2, options));
+    ASSERT_RAISES(Invalid, field2->MergeWith(field1, options));
+  }
```

Comment:
I think it would be nicer to also check a part of the error message. For example (untested, but you get the idea):
```suggestion
  void CheckUnifyFailsInvalid(
      const std::shared_ptr<Field>& field1, const std::shared_ptr<Field>& field2,
      const Field::MergeOptions& options = Field::MergeOptions::Defaults(),
      const std::string& match_message = "") {
    ARROW_SCOPED_TRACE("options: ", options);
    ARROW_SCOPED_TRACE("field2: ", field2->ToString());
    ARROW_SCOPED_TRACE("field1: ", field1->ToString());
    EXPECT_RAISES_WITH_MESSAGE_THAT(
        Invalid, ::testing::HasSubstr(match_message), field1->MergeWith(field2, options));
    EXPECT_RAISES_WITH_MESSAGE_THAT(
        Invalid, ::testing::HasSubstr(match_message), field2->MergeWith(field1, options));
  }
```